### PR TITLE
*: Optimize the logic of creating sqlrunner. #229

### LINKED ISSUE #229: [feature] Optimize the logic of creating sqlrunner.
**Is your feature request related to a problem? Please describe.**

Some parameters do not need to be retrieved every time sqlrunner is created.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by runkecheng**
### What type of PR is this?

Optimize the logic of creating sqlrunner.

### Which issue(s) this PR fixes?

Fixes #229

### What this PR does?

Summary: Apply the necessary configuration to SQLRunnerFactory to create the corresponding sqlrunner, which simplifies the creation process of sqlrunner.

### CODE REVIEW

File: utils/constants.go [hunk 1]
```diff
@@ -96,6 +98,9 @@ const (
 
 	// preUpdate file
 	FileIndicateUpdate = "PreUpdating"
+
+	// LeaderNode is the alias for leader`s host.
```
Comment (andyli029) [inline utils/constants.go:102]: Fix `LeaderNode `

File: internal/sql_runner.go [hunk 2]
```diff
@@ -38,30 +43,119 @@ var errorConnectionStates = []string{
...
+	db *sql.DB
+}
+
+// SQLRunner interface is a subset of mysql.DB.
+type SQLRunner interface {
+	RunQuery(query string, args ...interface{}) error
+	ClusterOperation
+}
+
+// ClusterOperation interface contains sql operations for cluster.
```
Comment (andyli029) [inline internal/sql_runner.go:112]: Fix the interface.

File: cluster/cluster.go [hunk 3]
```diff
@@ -331,3 +333,28 @@ func sizeToBytes(s string) (uint64, error) {
 	}
 	return 0, fmt.Errorf("'%s' format error, must be a positive integer with a unit of measurement like K, M or G", s)
 }
+
+// IsMysqlClusterKind for the given kind checks if CRD kind is for MysqlCluster CRD.
+func IsMysqlClusterKind(kind string) bool {
+	switch kind {
+	case "MysqlCluster", "mysqlcluster", "mysqlclusters":
```
Comment (acekingke) [inline cluster/cluster.go]: Kind do not have Mysql prefix

File: cluster/syncer/statefulset.go [hunk 4]
```diff
@@ -258,6 +262,13 @@ func (s *StatefulSetSyncer) updatePod(ctx context.Context) error {
 // 5. Check followerHost current role.
 // 6. If followerHost is not leader, switch it to leader through xenon.
 func (s *StatefulSetSyncer) preUpdate(ctx context.Context, leader, follower string) error {
+	sqlRunner, closeConn, err := s.SQLRunnerFactory(internal.NewConfigFromClusterKey(
```
Comment (acekingke) [inline cluster/syncer/statefulset.go:265]: sqlRunner should change name to leaderRunner， indicating that is used for leader

File: internal/query.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+package internal
+
+import (
+	"strings"
+	"errors"
+)
+
+// Query contains a escaped query string with variables marked with a question mark (?) and a slice
+// of positional arguments
```
Comment (andyli029) [inline internal/query.go:25]: Add `.` to the end of the line.

File: internal/sql_runner.go [hunk 6]
```diff
@@ -38,36 +43,141 @@ var errorConnectionStates = []string{
...
+type SQLRunnerFactory func(cfg *Config, errs ...error) (SQLRunner, func(), error)
+
+// NewSQLRunner opens a connections using the given DSN.
+func NewSQLRunner(cfg *Config, errs ...error) (SQLRunner, func(), error) {
+	var db *sql.DB
+	var close func()
+
+	// Make this factory accept a functions that tries to generate a config.
+	if len(errs) > 0 && errs[0] != nil {
+		return nil, close, errs[0]
```
Comment (acekingke) [inline internal/sql_runner.go:133]: return close is correct ？

File: cluster/cluster.go [hunk 7]
```diff
@@ -331,3 +333,28 @@ func sizeToBytes(s string) (uint64, error) {
 	}
 	return 0, fmt.Errorf("'%s' format error, must be a positive integer with a unit of measurement like K, M or G", s)
 }
+
+// IsMysqlClusterKind for the given kind checks if CRD kind is for Cluster CRD.
+func IsMysqlClusterKind(kind string) bool {
```
Comment (acekingke) [inline cluster/cluster.go:338]: funciton Name need change? where it be called？

File: internal/sql_runner.go (hunk 6 above)
Comment (runkecheng) [reply internal/sql_runner.go:133]: In this position, close and nil are equivalent, I think close can be used to return.

File: cluster/cluster.go (hunk 7 above)
Comment (runkecheng) [reply cluster/cluster.go:338]: This function is not used in pr this time, I will remove it.

File: internal/sql_runner.go [hunk 8]
```diff
@@ -38,36 +43,141 @@ var errorConnectionStates = []string{
...
+	QueryRows(query Query) (*sql.Rows, error)
+}
+
+// SQLRunnerFactory a function that generates a new SQLRunner.
+type SQLRunnerFactory func(cfg *Config, errs ...error) (SQLRunner, func(), error)
+
+// NewSQLRunner opens a connections using the given DSN.
+func NewSQLRunner(cfg *Config, errs ...error) (SQLRunner, func(), error) {
+	var db *sql.DB
+	var close func()
```
Comment (acekingke) [inline internal/sql_runner.go:129]: at best initial with nil

**Comment by acekingke** [comment]
LGTM

**Review by andyli029 — COMMENTED** [review]
LGTM.

**Pull request merged**